src/Microsoft.Identity.Client/Internal/ClientCredentialWrapper.cs

```diff
@@ -17,41 +24,154 @@ namespace Microsoft.Identity.Client.Internal
+                break;
+            case ConfidentialClientAuthenticationType.ClientCertificateWithClaims:
+                Certificate = config.ClientCredentialCertificate;
+                ClaimsToSign = config.ClaimsToSign;
+                break;
+            case ConfidentialClientAuthenticationType.ClientSecret:
+                Secret = config.ClientSecret;
+                break;
+            case ConfidentialClientAuthenticationType.SignedClientAssertion:
+                SignedAssertion = config.SignedClientAssertion;
+                break;
+            }
         }
 
-        internal ClientAssertionCertificateWrapper Certificate { get; private set; }
-        internal string Assertion { get; set; }
-        internal long ValidTo { get; set; }
-        internal bool ContainsX5C { get; set; }
-        internal string Audience { get; set; }
+        #region TestConstructors
+        //The following constructors are inteded for testing
+        public static ClientCredentialWrapper CreateWithCertificate(X509Certificate2 certificate)
+        {
+            return new ClientCredentialWrapper(certificate);
+        }
+
+        public static ClientCredentialWrapper CreateWithSecret(string secret)
+        {
+            return new ClientCredentialWrapper(secret, ConfidentialClientAuthenticationType.ClientSecret);
+        }
+
+        public static ClientCredentialWrapper CreateWithSignedClientAssertion(string signedClientAssertion)
+        {
+            return new ClientCredentialWrapper(signedClientAssertion, ConfidentialClientAuthenticationType.SignedClientAssertion);
+        }
 
-        /// <summary>
-        /// Constructor of client (application) credentials from a client secret, also known as the application password.
-        /// </summary>
-        /// <param name="secret">Secret string previously shared with AAD at application registration to prove the identity
-        /// of the application (the client) requesting the tokens.</param>
-        public ClientCredentialWrapper(string secret)
+        private ClientCredentialWrapper(X509Certificate2 certificate)
         {
             ConfidentialClientApplication.GuardMobileFrameworks();
 
-            if (string.IsNullOrWhiteSpace(secret))
+            Certificate = certificate;
+        }
+
+        private ClientCredentialWrapper(string secretOrAssertion, ConfidentialClientAuthenticationType authType)
+        {
+            ConfidentialClientApplication.GuardMobileFrameworks();
+
+            if (authType == ConfidentialClientAuthenticationType.SignedClientAssertion)
+            {
+                SignedAssertion = secretOrAssertion;
+            }
+            else
             {
-                throw new ArgumentNullException(nameof(secret));
+                Secret = secretOrAssertion;
             }
+        }
+        #endregion TestConstructors
+        private void CheckCertificateKeySize(X509Certificate2 cert)
+        {
+#if DESKTOP
```

Comment:
Please add a comment on why we only check this on DESKTOP and not on NETCORE and NETSTANDARD (i.e. breaking changes problems etc.). Perhaps also reference a github issue.

Reply:
This comment is important, as suggested I would prefer a link to the GitHub issue. Also we should probably have a min size regardless of target

---
In reply to: [295260535](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1208#discussion_r295260535) [](ancestors = 295260535)